source/host_resolver.c

```diff
@@ -311,25 +320,88 @@ static void s_remove_host_listener_from_entry(
 
 static void s_host_listener_destroy(struct host_listener *listener);
 
+struct host_purge_callback_options {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+    aws_simple_completion_callback *on_purge_cache_complete_callback;
+    void *user_data;
+};
+
+static void s_host_purge_callback_options_destroy(void *user_data) {
+    struct host_purge_callback_options *options = user_data;
+    options->on_purge_cache_complete_callback(options->user_data);
+    aws_mem_release(options->allocator, options);
+}
+
+static struct host_purge_callback_options *s_host_purge_callback_options_new(
+    struct aws_allocator *allocator,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+    struct host_purge_callback_options *purge_callback_options =
+        aws_mem_calloc(allocator, 1, sizeof(struct host_purge_callback_options));
+    purge_callback_options->allocator = allocator;
+    aws_ref_count_init(
+        &purge_callback_options->ref_count, purge_callback_options, s_host_purge_callback_options_destroy);
+    purge_callback_options->on_purge_cache_complete_callback = on_purge_cache_complete_callback;
+    purge_callback_options->user_data = user_data;
+    return purge_callback_options;
+}
+
+static void s_purge_cache_callback(void *user_data) {
+    struct host_purge_callback_options *purge_callback_options = user_data;
+    aws_ref_count_release(&purge_callback_options->ref_count);
+}
+
 /*
  * resolver lock must be held before calling this function
  */
-static void s_clear_default_resolver_entry_table(struct default_host_resolver *resolver) {
+static void s_clear_default_resolver_entry_table(
+    struct default_host_resolver *resolver,
+    aws_simple_completion_callback *on_purge_cache_complete_callback,
+    void *user_data) {
+
+    struct host_purge_callback_options *purge_callback_options = NULL;
+    if (on_purge_cache_complete_callback) {
+        purge_callback_options =
+            s_host_purge_callback_options_new(resolver->allocator, on_purge_cache_complete_callback, user_data);
+    }
+
     struct aws_hash_table *table = &resolver->host_entry_table;
     for (struct aws_hash_iter iter = aws_hash_iter_begin(table); !aws_hash_iter_done(&iter);
          aws_hash_iter_next(&iter)) {
         struct host_entry *entry = iter.element.value;
+        if (on_purge_cache_complete_callback) {
+            /* acquire a refernce to wait for the callback to trigger */
+            aws_ref_count_acquire(&purge_callback_options->ref_count);
+            entry->on_host_purge_complete = s_purge_cache_callback;
+            entry->on_host_purge_complete_user_data = purge_callback_options;
+        }
         s_shutdown_host_entry(entry);
```

Comment:
I think this is not worth to be a separate call. Always freak me out when we held the lock and call into other functions. And this function is just too simple and only invoked once.

Reply:
Thanks, fixed.